Introduction to Free Monads/article.md

```diff
@@ -0,0 +1,702 @@
+# Introduction to Free Monads
+
+If you've been around Haskell circles for a bit, you've probably seen the term "free monads".
+If you're relatively new to Haskell, you might feel that monads are hard enough to grasp as is, and now there's these different type of monads to worry about.
+Perhaps you've used Haskell for a while but didn't yet get to wrapping your head around free monads.
+In any case, the concept is ultimately rather straightforward.
+This article aims to introduce free monads and to explain why they are useful.
+
+Basic familiarity with Haskell is assumed for the rest of this article.
+Specifically, the `do`-notation, and type classes `Monoid`, `Functor`, and `Monad`.
+
+## Free algebraic structures
+
+> To understand the concept, you should think of "free" as in "free speech," not as in "free beer".
+>
+> -- Free Software Foundation, What is Free Software?
+
+So, to start with, let us try to explore what "free" in free monads stands for.
+This part is a bit theory-heavy, so if it is not your cup of tea, feel free to skim over it.
+
+In abstract algebra, we call something a "free X" when it is in some sense a minimal structure that satisfies conditions for being X.
+The exact definition of "minimal" in abstract algebra is a little vague, but the main idea is that "free X" satisfies all laws for X, and does not have any other laws describing it.
+That is, in this context, "free" means "unrestricted" -- it is not that we can get a structure "for free"[^for-free], but that no other laws are restricting the structure apart from those absolutely necessary.
+
+[^for-free]: Although in most cases, we actually might, depending on your definition of "for free".
+
+More formally, a free structure over a set $S$ is a set $M,$ together with operations on elements of $M,$ such that:
+
+- There is an embedding $i: S\to M$.
```

Comment:
For me, as a novice in abstract algebra, "embedding" wasn't clear from the first glance. I got it's some mapping, but then went to google it. So, it would be nice to have a relevant link here.

Moreover, is it really `S -> M`, not `M -> S`?

Reply:
Embedding is an injective morphism, I can just add a footnote explaining that. The point is every element in the original set is uniquely representable in the free structure, so it has to be `S → M` -- were it `M → S` there is no guarantee that we could uniquely (or at all) represent elements of the original set in the structure, which would make the claim that it's the structure over set `S` highly dubious.

IIRC below I give the example of a free monoid over type `T` being a list in Hask, where the embedding is `embed x = [x]` (i.e. essentially `pure`) -- the function isn't bijective mind, there's no injective inverse `unembed :: [a] -> a`, so that hopefully illustrates the point of embedding having to be `S → M` and not vice versa (here `S` being `a` ∀ a ∈ Hask and `M` being `[a]` ∀ a ∈ Hask)

Reply:
Oops, you've already added the footnote, 🆒